semantic-core/src/Core/Expr.hs

```diff
@@ -0,0 +1,51 @@
+{-# LANGUAGE OverloadedStrings #-}
+module Core.Expr
+( Expr(..)
+, id'
+, const'
+, true
+, false
+, not'
+, and'
+, or'
+, ($$)
+, Name
+) where
+
+import Analysis.Name
+
+-- | A lambda calculus with integers, conditionals, and exceptions.
+data Expr
+  = Var Name
+  | Abs Name Expr
+  | App Expr Expr
+  | Let Name Expr Expr
+  | Lit Int
+  | If Expr Expr Expr
+  | Die String
+  deriving (Eq, Ord, Show)
```

Comment:
This is a minimalistic reference calculus that exercises the evaluation interfaces fully.